### PR TITLE
crimson/os/seastore/async_cleaner: factor out journal trimming implementation

### PR DESCRIPTION
**Pull request by cyx1231st**
This is to make trimming and segment-reclaim loosely coupled with composition relationship, as the cold device tier will not need trimming.

The next step will be to move `GCProcess` up to `EPM` with proper refactors.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
jenkins retest this please

**Comment by cyx1231st** [comment]
jenkins retest this please

**Comment by cyx1231st** [comment]
Might be related to @zhscn

**Comment by cyx1231st** [comment]
jenkins test make check

**Comment by zhscn** [comment]
LGTM. 

I'm working on evicting the cold data to HDD device, I will open a PR after this refactor is merged

**Comment by cyx1231st** [comment]
> The next step will be to move GCProcess up to EPM with proper refactors.

... and to fully decouple trimming and segment cleaning.

@zhscn This is the direction I'm currently working on. I think the architecture with EPM/Trimmer/Cleaner would be much cleaner after that.

**Comment by zhscn** [comment]
> ... and to fully decouple trimming and segment cleaning.

Do you mean I should wait you to finish decoupling?

**Comment by cyx1231st** [comment]
> @cyx1231st Do you mean I should wait you to finish decoupling?

There will be no major changes to the existing logic (I suppose), mostly about the abstraction work and to move logics accordingly. As far as I understand, the major impacts would be:
* `AsyncCleaner::GCProcess` to become `EPM::BackgroundProcess` that manages a `trimmer` and a tier-0 `cleaner` as a start. There will be some work to reverse the dependencies.
* Introduce a segment-independent `AsyncCleaner` interface to `EPM::BackgroundProcess`, and rename the current concrete class back to `SegmentCleaner`. I think it will also unblock @myoungwon's work to introduce `RBCleaner` as an alternative implementation.

If this direction makes sense, I think it is worth prototyping how to move cold data to HDD device at the same time.

@zhscn what do you think?

File: src/crimson/os/seastore/async_cleaner.h
Comment (myoungwon) [inline src/crimson/os/seastore/async_cleaner.h:1183]: I don't see any function or place where this blocking_background is set. Am I correct?

**Review by myoungwon — APPROVED** [review]

**Review by xxhdx1985126 — APPROVED** [review]

**Review by athanatos — APPROVED** [review]

**Pull request merged**
